### ISSUE TITLE
Garbage 0x80 bytes on standby...

### ISSUE THREAD

**Issue by Kliodna** [jchristn/CavemanTcp#2]
Hello. After three days of unsuccessful struggle with the problem, I decided to write to you.
I would be very grateful if you can help me ...

I am using your library to connect to a board I developed based on STM32 controller. I used to start with NET CORE 3.1 and your library worked just like a charm. However, after switching to FRAMEWORK 4.8 in standby mode, the receiving buffer is filled with some periodicity with 0x80 trash bytes.

![Screenshot](https://user-images.githubusercontent.com/71749726/93972810-b05e5b80-fd7b-11ea-9e9d-ff574a8a45ae.png)

In the picture I am sending 2 bytes for a request and should receive 20 bytes as a response from the server.
The source code of the two applications is identical and does not differ by a single character.
The same problem is observed in your other library - SimpleTcp.

Can you somehow help me with this problem?
Thank you in advance...

**Comment by jchristn** [comment]
Hi @Kliodna wow, I can only imagine the amount of frustration this must be causing!  How is the board physically connected?  Is there an actual network in between your PC and the board?  Or is it Ethernet over USB or something to that effect?

If there's an actual network, I'd suggest taking two packet captures using Wireshark (in two separate files).  I've never seen an issue like this with my libraries (not to say that they couldn't be causing it, they certainly could be).  I'd be happy to take a look at the captures to see.

In the meantime, please also try disabling TCP keepalives prior to issuing the ```client.Connect(n)``` command and let me know if that changes anything (I doubt it will)

**Comment by Kliodna** [comment]
Hi again! Thank you very much for the quick response.
This morning I decided to try several options before installing Wireshark ...

1. tcpClient.Keepalive.EnableTcpKeepAlives = false; before tcpClient.Connect(1);
Strangely enough, it has helped to solve the problem and more 0x80 bytes do not come again.
However, now after physically disconnecting the board from the computer, the event Events.ClientDisconnected don't fire anymore.

2. I installed an older version of the library, namely 1.2.1.3
I had to change the code a bit, but in this case it also worked as expected. I did this because some time ago I already used your library in one of my home projects and there was no such problem. That is why I decided that the matter was in the library and its updates and decided to contact you.

Basically, now I can just use the old version of the library and that will be enough for me, but in my life I have a great desire to always use the latest versions ))) If you need more information from me to find and fix this problem, please let me know. I will gladly provide you with all the information you need...

**Comment by jchristn** [comment]
Hi @Kliodna thanks for the update.  The older version (1.2.1.3) did not have support for TCP keepalives (added in 1.3.0, here: https://github.com/jchristn/CavemanTcp/commit/e2f65053754f59826792bf6cf8a2b53b2371a027).  It seems that perhaps the controller on the system with which you are interacting doesn't like TCP keepalives.  

With 1.2.1.3, does the ClientDisconnect event fire as expected?  Thanks for getting back to me!

**Comment by Kliodna** [comment]
Yes, with version 1.2.1.3 everything works as it should.

**Comment by Kliodna** [comment]
But here's another interesting question. 
Why everything works with .net core 3.1 as it should, even without disabling TcpKeepAlives? ))

**Comment by jchristn** [comment]
Hi @Kliodna that's a great question.  CavemanTcp sits on top of TCP and is really just a wrapper for the APIs exposed by the underlying framework (and operating system) to make them easier to use.  Each is a little different, even though they are not supposed to be.  For instance, .NET Core provides full configurability of keepalive parameters.  .NET Standard doesn't provide the functionality, and .NET Framework provides less configurability and through different interfaces.  There are subtle differences in the underlying frameworks that make it really challenging to understand 'why' something works with this framework but not over here with this framework.  *Especially* in the area of disconnect detection (I have an open thread on this on another TCP library right now)

That doesn't explain the difference in behavior between 1.2.1.3 and 1.3.x, though, where aside from a minor refactor, the only change was adding (and enabling by default TCP keepalives).  As you rightly pointed out it should behave exactly the same in 1.2.1.3 as 1.3.x should with keepalives disabled.  I'm going to diff my commits to see if anything stands out on that right now.

Cheers, Joel

**Comment by jchristn** [comment]
Changes between 1.2.1.3 and 1.3.x (commit: https://github.com/jchristn/CavemanTcp/commit/e2f65053754f59826792bf6cf8a2b53b2371a027) 
- Class rename to avoid confusion with underlying TcpClient class
- Getters and setters for client/server settings, events, keepalive settings
- Rename of internal/private properties for clarity
- Moved client initialization from constructors (redundant) to common initializeclient method
- Added TCP keepalive initialization if enabled (they are enabled by default)

Out of curiosity, do you know much about the TCP implementation on the server-side on the board with which you are interacting?

Cheers, Joel
-

**Comment by jchristn** [comment]
Hi @Kliodna I haven't been able to find any difference in testing on my end with .NET Framework.  Bringing back up the last question in my previous comment, do you know much about the TCP implementation on the server-side on the board with which you are interacting?

**Comment by Kliodna** [comment]
Hello. I apologize for the long response time ...
Unfortunately, I do not really understand the details of TCP implementation. This is why I have to use third party libraries from more educated people like you ))) I am just at the beginning of my journey in this area.
Now your library works just fine and it suits me completely for my purposes. 
Thank you so much for your time. I think the topic can be closed.
